### PR TITLE
[#4] [API] As a user, I can access some JSON data

### LINKED ISSUE #4: [API] As a user, I can access some JSON data
## Why

Set up JSON endpoint for the APIs

## Acceptance Criteria

- Create new API module
- Create endpoint `GET` `/keywords`
    - Endpoint returns JSON with `JSONAPI`
```
[
		{
			"id": "1",
			"type": "keyword",
			"attributes": {
				"name": "First keyword"
			}
		},
		{
			"id": "1",
			"type": "keyword",
			"attributes": {
				"name": "Second keyword"
			}
		}
	]
```

### PR DESCRIPTION
**Pull request by blyscuit**
- Close #4 

## What happened 👀

Add API endpoint returning JSONAPI. 

## Insight 📝

- Add `jsonapi-serializer`.
- Add sample controller `KeywordController`.
- Coverage now working correctly by following the one line change from this article https://medium.com/geekculture/how-to-fix-inaccurate-test-coverage-outputs-from-simplecov-f4db553e7bee.

## Proof Of Work 📹

Screen Shot 2023-06-01 at 16 11 30

### CODE REVIEW

File: app/controllers/api/v1/keywords_controller.rb
Comment (blyscuit) [reply app/controllers/api/v1/keywords_controller.rb]: This will be removed later so can ignore.

File: spec/models/keyword_spec.rb
Comment (byhbt) [inline spec/models/keyword_spec.rb]:
```suggestion
```

Same with the comment here: https://github.com/blyscuit/rails-ic/pull/32/files#r1217354441

File: spec/requests/api/v1/keywords_controller_spec.rb
Comment (byhbt) [inline spec/requests/api/v1/keywords_controller_spec.rb]: We can use different verb to make the sentence more concise.

```suggestion
    it 'responds the valid keyword' do
```

File: app/controllers/api/v1/application_controller.rb
Comment (malparty) [inline app/controllers/api/v1/application_controller.rb]:
```suggestion
    class ApplicationController < ActionController::API; end
```

Since ruby 3 we can inline empty classes :) it doesn't change much, but just to let you know about it

File: app/controllers/api/v1/keywords_controller.rb
Comment (malparty) [inline app/controllers/api/v1/keywords_controller.rb]: As this is an `index` action, should we implement the right format directly? 
If yes, we can just serialize a list of keywords:

```suggestion
        keywords = [Keyword.new(1, 'First keyword'), Keyword.new(1, 'Second keyword')]
        
        render json: KeywordSerializer.new(keywords).serializable_hash.to_json
```

Comment (malparty) [reply app/controllers/api/v1/keywords_controller.rb]: Another option is to use the `show` action for this story, and to keep single item serialization ;-)

File: app/models/keyword.rb
Comment (malparty) [inline app/models/keyword.rb]:
```suggestion
  attr_accessor :id, :name
```

Does it work with just `attr_accessor`? (i.e. getter and setter)

File: config/environments/test.rb
Comment (malparty) [inline config/environments/test.rb]: Can you provide more insight why this change?

In general, we prefer to keep `ENV["CI"].present?` because:
- In local test run, we want tests to run as fast as possible, so not loading the whole code of the application allow to only load what the tests required and run faster
- In CI, we generally run the whole test suite, so there is no time gained by using lazy loading. However, using `eager loading` will ensure to raise an error/exception in case a part of the code is corrupted (even if there is no test running on that part of the code). So it allow to identify an error BEFORE the error is raised in the staging/production environment 🙌 

For that reason, if you had an issue with this, we might want to fix the issue and keep eager loading for CI 🤔 (did you add a `CI` env variable in your GitHub action env variables?)

File: spec/fabricators/keyword_fabricator.rb
Comment (malparty) [inline spec/fabricators/keyword_fabricator.rb]:
```suggestion
  id { init_with(FFaker::Identification.ssn }
  name { FFaker::FreedomIpsum.word) }
```

If you use the attr_accessor as suggested above, then you can use a simpler and more explicit Fabricator ;-)

File: spec/models/keyword_spec.rb
Comment (malparty) [inline spec/models/keyword_spec.rb]: I would not recommend testing the attributes of an object.
The main reason is that attributes are a mean to achieve something, but they are not an end. 
It is a bit similar for the "model associations", we could write beautiful tests to assert the one-to-many or many-to-many associations between models, but the tests would be similar to the code in the model (see [Tautological Tests](https://randycoulman.com/blog/2016/12/20/tautological-tests/)), so it is better to have tests that focus on the logic and that relies on these associations to test some business logic.


In case you really want to test it, I would recommend going 1 step further to be closer to testing business logic:

```suggestion
    it 'exposes its name' do
      keyword = Fabricate(:keyword, name: 'A keyword name')
      
      expect(keyword.name).to eq('A keyword name')
    end
```

but again, my opinion is that it is not needed 😇

**Review by malparty — CHANGES_REQUESTED** [review]

File: app/models/keyword.rb
Comment (blyscuit) [reply app/models/keyword.rb]: It works with `attr_accessor` but lint catches this because we are not setting the values, so `attr_reader` to limit the write access to these values sound better for me.

File: spec/fabricators/keyword_fabricator.rb
Comment (blyscuit) [reply spec/fabricators/keyword_fabricator.rb]: I prefer to keep using `attr_reader` for this to satisfy the line, I can come back to look at this if this model requires change to `attr_accessor` 🙏

File: spec/models/keyword_spec.rb
Comment (blyscuit) [reply spec/models/keyword_spec.rb]: I agree with you, this test is redundant of the init functionality of the object but I would like to keep it here so that the `keyword` model has some test that I can build on in the future features. I think of it as a starting point for other tests for this model.

Taking you suggestion, I will update the test to not use Fabricator for the `expose` test, as we are testing the `init` function and also the `read` permission.

File: spec/models/keyword_spec.rb
Comment (blyscuit) [reply spec/models/keyword_spec.rb]: Thank you very much, it seems I updated this in the wrong PR.

File: app/controllers/api/v1/keywords_controller.rb
Comment (blyscuit) [reply app/controllers/api/v1/keywords_controller.rb]: Sounds good, we can have it in the future-correct format. I will update the story.

File: config/environments/test.rb
Comment (blyscuit) [reply config/environments/test.rb]: Thank you for the detail explanation. I think I found some weird issues when using eager load on GH Actions but now it is resolved. Reverting this code back. 🙏

File: app/models/keyword.rb
Comment (malparty) [reply app/models/keyword.rb]: Lint is here to help us pay attention to the right things. But it is not here to lower the quality of the code or to make Fabricator more complex than what they should be 😇 

`keywords` aims to be an ActiveRecord model in the future. To have the least need for changes, `attr_accessor` will be used by the active record and therefore the way all other libraries are optimized (e.g. Fabrication).

While it's `working` with attr_reader and an initializer, it makes tests less standard and it will require you to refactor the code more later. That's the reason why I recommend using `attr_accessor` in this case.

File: spec/models/keyword_spec.rb
Comment (malparty) [reply spec/models/keyword_spec.rb]: Ok. 

Small note: `read permission` => It's usually called `visibility` (attributes visibility, method visibility, ... Either public, protected or private).

Let's keep `permission` for requests params and authorization ;-)

File: db/schema.rb
Comment (malparty) [inline db/schema.rb:1]: Before merging this PR, please remove this change 🙏 

The schema file is autogenerated, but its changes must always reflect the DB migrations added in the same PR. These changes come from another PR. 😇

File: spec/requests/api/v1/keywords_controller_spec.rb
Comment (malparty) [inline spec/requests/api/v1/keywords_controller_spec.rb]:
```suggestion
    it 'returns the success status' do
```

File: spec/requests/api/v1/keywords_controller_spec.rb
Comment (malparty) [inline spec/requests/api/v1/keywords_controller_spec.rb]:
```suggestion
    it 'returns the valid keywords' do
```

(for consistency with the previous test — both information come from the `response` object ;) )

**Review by malparty — APPROVED** [review]
Approved ahead with some small need for changes (mainly the schema file to rollback).

While I recommend using attr_accessor, it's not a blocker either and future refactoring might help better understand the way other libraries optimize for the `ActiveModel` code structure. So all good with this! :)

File: db/schema.rb
Comment (blyscuit) [reply db/schema.rb:1]: Good catch.

File: app/models/keyword.rb
Comment (blyscuit) [reply app/models/keyword.rb]: I see, thank you for the insight, I will update this model to `attr_accessor` to be future proof.

**Review by byhbt — APPROVED** [review]

File: app/models/keyword.rb
Comment (malparty) [inline app/models/keyword.rb:6]:
```suggestion
  include ActiveModel::Model

  attr_accessor :id, :name
```

Magic happens when you include `include ActiveModel::Model` 🚀 

You can then use it this way in the controller:

```
      def index
        first_keyword = Keyword.new({ id: 1, name: 'Test' })
        second_keyword = Keyword.new({ id: 2, name: 'Test' })
```

Comment (malparty) [reply app/models/keyword.rb:6]: Recommended reading about this: https://guides.rubyonrails.org/active_model_basics.html

Comment (blyscuit) [reply app/models/keyword.rb:6]: How do we decide between `ActiveModel::API` and `ActiveModel::Model`? From my understanding, Rail 7 forward, `ActiveModel::API` is preferred?

Comment (malparty) [reply app/models/keyword.rb:6]: It seems like currently they have the exact same behavior as ActiveModel includes ActiveAPI. But in the future, ActiveModel should have more "ActiveRecord like" features, while ActiveAPI aims to stay minimalist and focused on the View helpers (i.e. not including DB related helpers).

In this case, both are fine as in the future PR we will replace this `include ActiveModel/API` with a full ActiveReccord inheritance. ;-)

**Review by malparty — APPROVED** [review]

**Pull request merged**
